protocols/bftcosi/bftcosi.go

```diff
@@ -0,0 +1,615 @@
+	ch.Challenge = chal
+
+	dbg.Lvl3(bft.Name(), "BFTCoSi handle Challenge PREPARE")
+	// go to response if leaf
+	if bft.IsLeaf() {
+		return bft.startResponsePrepare()
+	}
+	var err error
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, ch)
+	}
+	return err
+}
+
+// handleCommitChallenge will verify the signature + check if no more than 1/3
+// of participants refused to sign.
+func (bft *ProtocolBFTCoSi) handleChallengeCommit(ch *ChallengeCommit) error {
+	ch.Challenge = bft.commit.Challenge(ch.Challenge)
+	prep := &msgPrepare{Msg: bft.Msg}
+	h := prep.Hash()
+
+	// verify if the signature is correct
+	if err := cosi.VerifyCosiSignatureWithException(bft.suite,
+		bft.aggregatedPublic, h, ch.Signature,
+		ch.Exceptions); err != nil {
+		dbg.Error(bft.Name(), "Verification of the signature failed:", err)
+		bft.signRefusal = true
+	}
+
+	// Check if we have no more than 1/3 failed nodes
+	if len(ch.Exceptions) > int(bft.threshold) {
+		dbg.Errorf("More than 1/3 (%d/%d) refused to sign ! ABORT",
+			len(ch.Exceptions), len(bft.EntityList().List))
+		bft.signRefusal = true
+	}
+
+	// store the exceptions for later usage
+	bft.tempExceptions = ch.Exceptions
+	dbg.Lvl3("BFTCoSi handle Challenge COMMIT")
+	if bft.IsLeaf() {
+		return bft.startResponseCommit()
+	}
+
+	// send it down
+	for _, tn := range bft.Children() {
+		if err := bft.SendTo(tn, ch); err != nil {
+			dbg.Error(err)
+		}
+	}
+	return nil
+}
+
+// startPrepareResponse wait the verification of the block and then start the
+// challenge process
+func (bft *ProtocolBFTCoSi) startResponsePrepare() error {
+	// create response
+	resp, err := bft.prepare.CreateResponse()
+	if err != nil {
+		return err
+	}
+	// wait the verification
+	r, ok := bft.waitResponseVerification()
+	if ok {
+		// append response only if OK
+		r.Response = resp
+	}
+	dbg.Lvl3(bft.Name(), "BFTCoSi Start Response PREPARE with response:", r)
+	// send to parent
+	return bft.SendTo(bft.Parent(), r)
+}
+
+// startCommitResponse will create the response for the commit phase and send it
+// up. It will not create the response if it decided the signature is wrong from
+// the prepare phase.
+func (bft *ProtocolBFTCoSi) startResponseCommit() error {
+	r := &Response{
+		TYPE: RoundCommit,
+	}
+	// if i dont want to sign
+	if bft.signRefusal {
+		r.Exceptions = append(r.Exceptions, cosi.Exception{
+			Public:     bft.Public(),
+			Commitment: bft.commit.GetCommitment(),
+		})
+	} else {
+		// otherwise i create the response
+		resp, err := bft.commit.CreateResponse()
+		if err != nil {
+			return err
+		}
+		r.Response = resp
+	}
+	dbg.Lvl3(bft.Name(), "BFTCoSi Start Response COMMIT")
+	// send to parent
+	err := bft.SendTo(bft.Parent(), r)
+	bft.Done()
+	return err
+}
+
+// handleResponseCommit handles the responses for the commit round during the
+// response phase.
+func (bft *ProtocolBFTCoSi) handleResponseCommit(r *Response) error {
+	// check if we have enough
+	// FIXME possible data race
```

Comment:
Is that very difficult to fix? Either remove the possible data race or the *FIXME*.